File: CPP/ByTier/D/2063_D_Game_With_Triangles.cpp
// Problem: CF 2063 D - Game With Triangles
// https://codeforces.com/contest/2063/problem/D

/*
D. Game With Triangles
Algorithms/Techniques: Greedy, Sorting, Binary Search, Prefix Sums
time complexity: O((n + m) log(n + m)) per test case
space complexity: O(n + m)

This solution computes the maximum number of operations (kmax) that can be performed by choosing 3 non-collinear points from n points on y=0 and m points on y=2 to form triangles.
For each valid number of operations k from 1 to kmax, we maximize the total area using a greedy approach:
- Sort both arrays a and b
- For k operations, try all possible distributions of chosen points between a and b to maximize triangle areas
- Use prefix sums for efficient computation of potential gains
- Binary search is used to find the optimal split of points for maximizing score in each operation count k.
*/

#include <algorithm>
#include <iostream>

using namespace std;
typedef long long ll;
const int N = 2e5 + 5;
int n, m, a[N], b[N];
ll sa[N], sb[N];

// Calculates the combined score given x points from array a and (k-x) from array b
ll calc(int x, int k) { return sa[x] + sb[k - x]; }

void clr() {}

void solve() {
  cin >> n >> m;
  for (int i = 1; i <= n; i++)
    cin >> a[i];
  for (int i = 1; i <= m; i++)
    cin >> b[i];
  sort(a + 1, a + n + 1), sort(b + 1, b + m + 1);
  
  // Determine maximum number of valid operations based on greedy selection
  int na = 0, nb = 0, kmax = 0;
  while (2 * (na + 1) <= n - nb && 2 * nb < m - na ||
         2 * (nb + 1) <= m - na && 2 * na < n - nb) {
    if (n - nb - 2 * na >= m - na - 2 * nb)
      kmax++, na++;
    else
      kmax++, nb++;
  }
  
  cout << kmax << '\n';
  
  // Compute prefix sums for areas contributed by choosing pairs from a and b.
  // sa[i] = sum of differences in sorted a for first i pairs (largest to smallest)
  for (int i = 1; i <= n / 2; i++)
    sa[i] = sa[i - 1] + a[n - i + 1] - a[i];
  for (int i = 1; i <= m / 2; i++)
    sb[i] = sb[i - 1] + b[m - i + 1] - b[i];

  // For each operation count k from 1 to kmax, find best way to distribute selections
  for (int k = 1; k <= kmax; k++) {
    const int LL = k - min({k, m / 2, n, m - k}),
              RR = min({k, n / 2, m, n - k});
    int L = LL, R = RR + 1, mid;

    // Binary search to find the optimal split minimizing the loss in area
    while (L < R - 1) {
      mid = L + R >> 1;
      if (calc(mid, k) >= calc(mid - 1, k))
        L = mid;
      else
        R = mid;
    }
    
    cout << calc(L, k) << ' ';
  }
  cout << '\n';
  clr();
}

int main() {
  ios::sync_with_stdio(0);
  int T;
  cin >> T;
  while (T--)
    solve();
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2063/2063_D_Game_With_Triangles.py
# Problem: CF 2063 D - Game With Triangles
# https://codeforces.com/contest/2063/problem/D

"""
D. Game With Triangles

Purpose:
This code solves a problem where we are given n points on the line y=0 and m points on the line y=2.
We can form triangles using three non-collinear points, and each triangle contributes its area to our score.
The goal is to determine the maximum number of operations (k_max) and then calculate f(k), the maximum possible score
achievable by performing exactly k operations.

Algorithm:
- We are essentially trying to pick triplets of points such that one point is from y=0 and two from y=2,
  or one from y=2 and two from y=0, in a way that maximizes total area.
- Points on the same horizontal line cannot form valid triangles (collinear).
- So, for maximum score:
    * We must combine points from both lines to make triangles.
    * The idea is to sort both point sets, then pair up the farthest points from each end of the sorted lists
      to maximize base lengths (and thus areas).
    * We then greedily select combinations that give maximum gain at each step.

Time Complexity:
  O(n log n + m log m) per test case due to sorting.
  The rest is linear in n and m.

Space Complexity:
  O(n + m) for storing arrays a and b, plus auxiliary space for storing differences (a_l, b_l).
"""

from sys import stdin

inp = stdin.readline
t = int(inp())
for _ in range(t):
    n, m = map(int, inp().split())
    a = [int(x) for x in inp().split()]
    b = [int(x) for x in inp().split()]
    a.sort()
    b.sort()
    # Compute differences between farthest points from centers for both lists
    a_l = [a[-1 - i] - a[i] for i in range(n // 2)]
    b_l = [b[-1 - i] - b[i] for i in range(m // 2)]
    k_max = min((n + m) // 3, n, m)
    print(k_max)
    k = []
    count_a = 0
    count_b = 0
    ans = 0
    # Greedily build the answer by choosing optimal triangle combinations at each step
    for i in range(1, k_max + 1):
        if count_a > n - i:   # If we have too many a-points selected, reduce one
            ans += b_l[count_b]
            count_b += 1
            count_a -= 1
            ans -= a_l[count_a]
        if count_b > m - i:
            ans += a_l[count_a]
            count_a += 1
            count_b -= 1
            ans -= b_l[count_b]
        if count_a == n - i:   # All remaining a-points must be used, so add one from b
            ans += b_l[count_b]
            count_b += 1
        elif count_b == m - i:  # All remaining b-points must be used, so add one from a
            ans += a_l[count_a]
            count_a += 1
        elif a_l[count_a] > b_l[count_b]:   # Pick the higher area difference
            ans += a_l[count_a]
            count_a += 1
        else:
            ans += b_l[count_b]
            count_b += 1
        k.append(ans)
    if k_max > 0:
        print(*k)


# https://github.com/VaHiX/codeForces/